engineering-plans/0002-subgraph-schema-merging.md

```diff
@@ -0,0 +1,118 @@
+
+  <dt>Engineering Plan pull request</dt>
+  <dd><a href="https://github.com/graphprotocol/rfcs/pull/5">Engineering Plan PR</a></dd>
+
+  <dt>Obsoletes (if applicable)</dt>
+
+  <dt>Date of submission</dt>
+  <dd>2019-12-09</dd>
+
+  <dt>Date of approval</dt>
+  <dd>TBD</dd>
+
+  <dt>Approved by</dt>
+  <dd>TBD</dd>
+</dl>
+
+## Summary
+
+Subgraph composition allows a subgraph to import types from a foreign subgraph. Imports are designed to be loosely coupled throughout the deployment and indexing phases, and tightly coupled during query execution for the subgraph.
+
+To generate an API schema for the subgraph, the local schema needs to be merged with each of the foreign schemas that it imports. The merging process needs to take the following into account:
+
+1. An imported schema not being available on the graph-node
+2. An imported schema without a definition for the type which the local subgraph imports
+3. A merged and cached api schema for a subgraph imported by name is updated
+4. Ability to tell which subgraph/schema each type belongs to
+
+## Implementation
+
+There are two important parts to schema merging:
+
+1. Merging the schema
+
+2. Invalidating a cache entry when a subgraph's dependency, referenced by name, is updated and remerging that schema
+
+### Schema Merging
+
+Add a `merged_schema(&Schema, HashMap<SchemaReference, Arc<Schema>) -> Schema` function to the `graphql::schema` crate which will add each of the imported 
+types to the proviced document with a @subgraphId diretive denoting which subgraph the type came from. The `api_schema` function will add all the necessary types and fields for the imported types without any changes.
+
+Schema before calling `merged_schema`:
+
+```graphql
+type _Schema_
+	@import(
+		types: ["B"],
+		from: { id: "..." }
+	)
+
+type A @entity {
+	id: ID!
+	foo: B!
+}
+```
+
+Schema after calling `merged_schema`
+
+```graphql
+type A @entity @subgraphId("...") {
+	id: ID!
+	foo: B!
+}
+
+type B @entity @subgraphId("...") {
+	id: ID!
+	bar: String
+}
+```
+
+After the schema document is merged, the `api_schema` function will be called.
+
+### Cache Invalidation
+
+In the initial implementation, the cache will need to check if an entry needs to be updated when it is accessed. To determine whether an entry in the schema cache has been invalidated, each subgraph schema it imports by name needs to be queried in the store for its most current version. The schema cache will maintain a record of the versions used for merging in each entry, and if the version previously used to merge is no longer current that entry needs to be invalidated an remerged before responding to the cache access.
+
+A more performant invalidation solution would be to have the cache maintain a listener notifying it every time a subgraph's current version changes. Upon receiving the notification it would scan the schemas in the cache for those which need to be remerged.
```

Comment:
That's a good idea; and I agree this is future work. I wonder if we should add a "Future Work" section to engineering plans? What do you think?

Reply:
Yeah, I think it would be useful to make explicit. It's also a nudge to pare down your work to the essentials